packages/core/src/di/host_tag_name_token.ts

```diff
@@ -0,0 +1,55 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {RuntimeError, RuntimeErrorCode} from '../errors';
+import {TNodeType} from '../render3/interfaces/node';
+import {getCurrentTNode} from '../render3/state';
+
+import {InjectionToken} from './injection_token';
+import {InjectFlags} from './interface/injector';
+
+/**
+ * A token that can be used to inject the tag name of the host node.
+ *
+ * @usageNotes
+ * ### Injecting a tag name that is known to exist
+ * ```typescript
+ * @Directive()
+ * class MyDir {
+ *   tagName: string = inject(HOST_TAG_NAME);
+ * }
+ * ```
+ *
+ * ### Optionally injecting a tag name
+ * ```typescript
+ * @Directive()
+ * class MyDir {
+ *   tagName: string | null = inject(HOST_TAG_NAME, {optional: true});
+ * }
+ * ```
+ * @publicApi
+ */
+export const HOST_TAG_NAME = new InjectionToken<string>(ngDevMode ? 'HOST_TAG_NAME' : '');
+
+// HOST_TAG_NAME should be resolved at the current node, similar to e.g. ElementRef,
+// so we manually specify __NG_ELEMENT_ID__ here, instead of using a factory.
+// tslint:disable-next-line:no-toplevel-property-access
+(HOST_TAG_NAME as any).__NG_ELEMENT_ID__ = (flags: InjectFlags) => {
+  const tNode = getCurrentTNode()!;
+  if (tNode.type & TNodeType.Element) {
```

Comment:
We can also add a check here to provide a better error message in case tag name is injected outside of the supported context:

```suggestion
  const tNode = getCurrentTNode();
  
  if (ngDevMode && tNode === null) {
    throw new RuntimeError(...);
  }
  
  if (tNode.type & TNodeType.Element) {
```